insta/src/snapshot.rs

```diff
@@ -511,42 +596,107 @@ impl Snapshot {
 #[derive(Debug, Clone)]
-pub struct SnapshotContents {
+pub enum SnapshotContents {
+    Text(TextSnapshotContents),
+    Binary(BinarySnapshotContents),
+}
+
+// Could be Cow, but I think limited savings
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct TextSnapshotContents {
     contents: String,
-    pub kind: SnapshotKind,
+    pub kind: TextSnapshotKind,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct BinarySnapshotContents {
+    // This is in an `Rc` because we need to be able to clone this struct cheaply and the contents
+    // of the `Vec` could be rather large. The reason it's not an `Rc<[u8]>` is because creating one
+    // of those would require re-allocating because of the additional size needed for the reference
+    // count.
+    pub contents: Rc<Vec<u8>>,
+    pub extension: String,
+}
+
+impl From<TextSnapshotContents> for SnapshotContents {
+    fn from(value: TextSnapshotContents) -> Self {
+        SnapshotContents::Text(value)
+    }
+}
+
+impl From<BinarySnapshotContents> for SnapshotContents {
+    fn from(value: BinarySnapshotContents) -> Self {
+        SnapshotContents::Binary(value)
+    }
 }
 
 impl SnapshotContents {
-    pub fn new(contents: String, kind: SnapshotKind) -> SnapshotContents {
-        // We could store a normalized version of the string as part of `new`;
-        // it would avoid allocating a new `String` when we get the normalized
-        // versions, which we may do a few times. (We want to store the
-        // unnormalized version because it allows us to use `matches_fully`.)
-        SnapshotContents { contents, kind }
+    pub fn is_binary(&self) -> bool {
+        matches!(self, SnapshotContents::Binary(_))
     }
 
-    /// Returns the snapshot contents as string without any normalization
-    pub fn as_str_exact(&self) -> &str {
-        self.contents.as_str()
+    pub fn as_string_contents(&self) -> Option<&TextSnapshotContents> {
+        match self {
+            SnapshotContents::Text(contents) => Some(contents),
+            SnapshotContents::Binary(_) => None,
+        }
     }
+}
 
-    /// Matches another snapshot without any normalization
-    pub fn matches_fully(&self, other: &SnapshotContents) -> bool {
-        self.as_str_exact() == other.as_str_exact()
+impl TextSnapshotContents {
+    pub fn new(contents: String, kind: TextSnapshotKind) -> TextSnapshotContents {
+        // We could store a normalized version of the string as part of `new`;
+        // it would avoid allocating a new `String` when we get the normalized
+        // versions, which we may do a few times. (We want to store the
+        // unnormalized version because it allows us to use `matches_fully`.)
+        TextSnapshotContents { contents, kind }
     }
 
     /// Snapshot matches based on the latest format.
-    pub fn matches_latest(&self, other: &SnapshotContents) -> bool {
+    pub fn matches_latest(&self, other: &TextSnapshotContents) -> bool {
```

Comment:
(I should have done this before)
```suggestion
    pub fn matches_latest(&self, other: &Self) -> bool {
```